From da1bb145bae91c97d4696a64612af165f0f2f206 Mon Sep 17 00:00:00 2001
From: Daiki Ueno <ueno@gnu.org>
Date: Thu, 26 Feb 2026 07:50:42 +0900
Subject: [PATCH 01/11] BACKPORT: nettle: use SHA*_BLOCK_SIZE instead of
 deprecated SHA*_DATA_SIZE

Signed-off-by: Daiki Ueno <ueno@gnu.org>
(cherry picked from commit c86ce8cea337b2549e8b9e2387e0996f4fab198b)
Signed-off-by: Xi Ruoyao <xry111@xry111.site>
---
 lib/accelerated/aarch64/aarch64-common.h |  2 +-
 lib/accelerated/aarch64/sha-aarch64.c    | 24 ++++++++++++------------
 lib/accelerated/x86/hmac-padlock.c       | 20 ++++++++++----------
 lib/accelerated/x86/sha-padlock.c        |  4 ++--
 lib/accelerated/x86/sha-x86-ssse3.c      | 24 ++++++++++++------------
 lib/accelerated/x86/x86-common.h         |  2 +-
 6 files changed, 38 insertions(+), 38 deletions(-)

diff --git a/lib/accelerated/aarch64/aarch64-common.h b/lib/accelerated/aarch64/aarch64-common.h
index fde4fd937..8dfe236e0 100644
--- a/lib/accelerated/aarch64/aarch64-common.h
+++ b/lib/accelerated/aarch64/aarch64-common.h
@@ -28,7 +28,7 @@
 	{ #name,                                      \
 	  sizeof(struct name##_ctx),                  \
 	  NAME##_DIGEST_SIZE,                         \
-	  NAME##_DATA_SIZE,                           \
+	  NAME##_BLOCK_SIZE,                          \
 	  (nettle_hash_init_func *)name##_init,       \
 	  (nettle_hash_update_func *)update_func,     \
 	  (nettle_hash_digest_func *)digest_func }
diff --git a/lib/accelerated/aarch64/sha-aarch64.c b/lib/accelerated/aarch64/sha-aarch64.c
index 3b7089c80..c46e519d9 100644
--- a/lib/accelerated/aarch64/sha-aarch64.c
+++ b/lib/accelerated/aarch64/sha-aarch64.c
@@ -84,7 +84,7 @@ void aarch64_sha1_update(struct sha1_ctx *ctx, size_t length,
 	unsigned t2, i;
 
 	if ((res = ctx->index)) {
-		res = SHA1_DATA_SIZE - res;
+		res = SHA1_BLOCK_SIZE - res;
 		if (length < res)
 			res = length;
 		sha1_update(ctx, res, data);
@@ -98,14 +98,14 @@ void aarch64_sha1_update(struct sha1_ctx *ctx, size_t length,
 	octx.h3 = ctx->state[3];
 	octx.h4 = ctx->state[4];
 
-	memcpy(octx.data, ctx->block, SHA1_DATA_SIZE);
+	memcpy(octx.data, ctx->block, SHA1_BLOCK_SIZE);
 	octx.num = ctx->index;
 
-	res = length % SHA1_DATA_SIZE;
+	res = length % SHA1_BLOCK_SIZE;
 	length -= res;
 
 	if (length > 0) {
-		t2 = length / SHA1_DATA_SIZE;
+		t2 = length / SHA1_BLOCK_SIZE;
 
 		sha1_block_data_order(&octx, data, t2);
 
@@ -142,7 +142,7 @@ void aarch64_sha256_update(struct sha256_ctx *ctx, size_t length,
 	unsigned t2, i;
 
 	if ((res = ctx->index)) {
-		res = SHA256_DATA_SIZE - res;
+		res = SHA256_BLOCK_SIZE - res;
 		if (length < res)
 			res = length;
 		sha256_update(ctx, res, data);
@@ -151,14 +151,14 @@ void aarch64_sha256_update(struct sha256_ctx *ctx, size_t length,
 	}
 
 	memcpy(octx.h, ctx->state, sizeof(octx.h));
-	memcpy(octx.data, ctx->block, SHA256_DATA_SIZE);
+	memcpy(octx.data, ctx->block, SHA256_BLOCK_SIZE);
 	octx.num = ctx->index;
 
-	res = length % SHA256_DATA_SIZE;
+	res = length % SHA256_BLOCK_SIZE;
 	length -= res;
 
 	if (length > 0) {
-		t2 = length / SHA1_DATA_SIZE;
+		t2 = length / SHA1_BLOCK_SIZE;
 		sha256_block_data_order(&octx, data, t2);
 
 		for (i = 0; i < t2; i++)
@@ -193,7 +193,7 @@ void aarch64_sha512_update(struct sha512_ctx *ctx, size_t length,
 	unsigned t2, i;
 
 	if ((res = ctx->index)) {
-		res = SHA512_DATA_SIZE - res;
+		res = SHA512_BLOCK_SIZE - res;
 		if (length < res)
 			res = length;
 		sha512_update(ctx, res, data);
@@ -202,14 +202,14 @@ void aarch64_sha512_update(struct sha512_ctx *ctx, size_t length,
 	}
 
 	memcpy(octx.h, ctx->state, sizeof(octx.h));
-	memcpy(octx.u.p, ctx->block, SHA512_DATA_SIZE);
+	memcpy(octx.u.p, ctx->block, SHA512_BLOCK_SIZE);
 	octx.num = ctx->index;
 
-	res = length % SHA512_DATA_SIZE;
+	res = length % SHA512_BLOCK_SIZE;
 	length -= res;
 
 	if (length > 0) {
-		t2 = length / SHA512_DATA_SIZE;
+		t2 = length / SHA512_BLOCK_SIZE;
 		sha512_block_data_order(&octx, data, t2);
 
 		for (i = 0; i < t2; i++)
diff --git a/lib/accelerated/x86/hmac-padlock.c b/lib/accelerated/x86/hmac-padlock.c
index 8da42b327..2c068372d 100644
--- a/lib/accelerated/x86/hmac-padlock.c
+++ b/lib/accelerated/x86/hmac-padlock.c
@@ -280,39 +280,39 @@ static int wrap_padlock_hmac_fast(gnutls_mac_algorithm_t algo,
 {
 	if (algo == GNUTLS_MAC_SHA1 || algo == GNUTLS_MAC_SHA256) {
 		unsigned char *pad;
-		unsigned char pad2[SHA1_DATA_SIZE + MAX_SHA_DIGEST_SIZE];
+		unsigned char pad2[SHA1_BLOCK_SIZE + MAX_SHA_DIGEST_SIZE];
 		unsigned char hkey[MAX_SHA_DIGEST_SIZE];
 		unsigned int digest_size =
 			_gnutls_mac_get_algo_len(mac_to_entry(algo));
 
-		if (key_size > SHA1_DATA_SIZE) {
+		if (key_size > SHA1_BLOCK_SIZE) {
 			wrap_padlock_hash_fast((gnutls_digest_algorithm_t)algo,
 					       key, key_size, hkey);
 			key = hkey;
 			key_size = digest_size;
 		}
 
-		pad = gnutls_malloc(text_size + SHA1_DATA_SIZE);
+		pad = gnutls_malloc(text_size + SHA1_BLOCK_SIZE);
 		if (pad == NULL)
 			return gnutls_assert_val(GNUTLS_E_MEMORY_ERROR);
 
-		memset(pad, IPAD, SHA1_DATA_SIZE);
+		memset(pad, IPAD, SHA1_BLOCK_SIZE);
 		memxor(pad, key, key_size);
 
-		memcpy(&pad[SHA1_DATA_SIZE], text, text_size);
+		memcpy(&pad[SHA1_BLOCK_SIZE], text, text_size);
 
 		wrap_padlock_hash_fast((gnutls_digest_algorithm_t)algo, pad,
-				       text_size + SHA1_DATA_SIZE,
-				       &pad2[SHA1_DATA_SIZE]);
+				       text_size + SHA1_BLOCK_SIZE,
+				       &pad2[SHA1_BLOCK_SIZE]);
 
-		zeroize_key(pad, text_size + SHA1_DATA_SIZE);
+		zeroize_key(pad, text_size + SHA1_BLOCK_SIZE);
 		gnutls_free(pad);
 
-		memset(pad2, OPAD, SHA1_DATA_SIZE);
+		memset(pad2, OPAD, SHA1_BLOCK_SIZE);
 		memxor(pad2, key, key_size);
 
 		wrap_padlock_hash_fast((gnutls_digest_algorithm_t)algo, pad2,
-				       digest_size + SHA1_DATA_SIZE, digest);
+				       digest_size + SHA1_BLOCK_SIZE, digest);
 
 		zeroize_key(pad2, sizeof(pad2));
 		zeroize_key(hkey, sizeof(hkey));
diff --git a/lib/accelerated/x86/sha-padlock.c b/lib/accelerated/x86/sha-padlock.c
index edc2c63a0..2a4a4de45 100644
--- a/lib/accelerated/x86/sha-padlock.c
+++ b/lib/accelerated/x86/sha-padlock.c
@@ -189,8 +189,8 @@ static void padlock_sha512_digest(struct sha512_ctx *ctx, size_t length,
 	/* This is slightly inefficient, as the numbers are converted to
 	   big-endian format, and will be converted back by the compression
 	   function. It's probably not worth the effort to fix this. */
-	WRITE_UINT64(ctx->block + (SHA512_DATA_SIZE - 16), high);
-	WRITE_UINT64(ctx->block + (SHA512_DATA_SIZE - 8), low);
+	WRITE_UINT64(ctx->block + (SHA512_BLOCK_SIZE - 16), high);
+	WRITE_UINT64(ctx->block + (SHA512_BLOCK_SIZE - 8), low);
 	SHA512_COMPRESS(ctx, ctx->block);
 
 	words = length / 8;
diff --git a/lib/accelerated/x86/sha-x86-ssse3.c b/lib/accelerated/x86/sha-x86-ssse3.c
index 0f3e613b8..28d7930e6 100644
--- a/lib/accelerated/x86/sha-x86-ssse3.c
+++ b/lib/accelerated/x86/sha-x86-ssse3.c
@@ -82,7 +82,7 @@ void x86_sha1_update(struct sha1_ctx *ctx, size_t length, const uint8_t *data)
 	unsigned t2, i;
 
 	if ((res = ctx->index)) {
-		res = SHA1_DATA_SIZE - res;
+		res = SHA1_BLOCK_SIZE - res;
 		if (length < res)
 			res = length;
 		sha1_update(ctx, res, data);
@@ -96,14 +96,14 @@ void x86_sha1_update(struct sha1_ctx *ctx, size_t length, const uint8_t *data)
 	octx.h3 = ctx->state[3];
 	octx.h4 = ctx->state[4];
 
-	memcpy(octx.data, ctx->block, SHA1_DATA_SIZE);
+	memcpy(octx.data, ctx->block, SHA1_BLOCK_SIZE);
 	octx.num = ctx->index;
 
-	res = length % SHA1_DATA_SIZE;
+	res = length % SHA1_BLOCK_SIZE;
 	length -= res;
 
 	if (length > 0) {
-		t2 = length / SHA1_DATA_SIZE;
+		t2 = length / SHA1_BLOCK_SIZE;
 
 		sha1_block_data_order(&octx, data, t2);
 
@@ -140,7 +140,7 @@ void x86_sha256_update(struct sha256_ctx *ctx, size_t length,
 	unsigned t2, i;
 
 	if ((res = ctx->index)) {
-		res = SHA256_DATA_SIZE - res;
+		res = SHA256_BLOCK_SIZE - res;
 		if (length < res)
 			res = length;
 		sha256_update(ctx, res, data);
@@ -149,14 +149,14 @@ void x86_sha256_update(struct sha256_ctx *ctx, size_t length,
 	}
 
 	memcpy(octx.h, ctx->state, sizeof(octx.h));
-	memcpy(octx.data, ctx->block, SHA256_DATA_SIZE);
+	memcpy(octx.data, ctx->block, SHA256_BLOCK_SIZE);
 	octx.num = ctx->index;
 
-	res = length % SHA256_DATA_SIZE;
+	res = length % SHA256_BLOCK_SIZE;
 	length -= res;
 
 	if (length > 0) {
-		t2 = length / SHA1_DATA_SIZE;
+		t2 = length / SHA1_BLOCK_SIZE;
 		sha256_block_data_order(&octx, data, t2);
 
 		for (i = 0; i < t2; i++)
@@ -191,7 +191,7 @@ void x86_sha512_update(struct sha512_ctx *ctx, size_t length,
 	unsigned t2, i;
 
 	if ((res = ctx->index)) {
-		res = SHA512_DATA_SIZE - res;
+		res = SHA512_BLOCK_SIZE - res;
 		if (length < res)
 			res = length;
 		sha512_update(ctx, res, data);
@@ -200,14 +200,14 @@ void x86_sha512_update(struct sha512_ctx *ctx, size_t length,
 	}
 
 	memcpy(octx.h, ctx->state, sizeof(octx.h));
-	memcpy(octx.u.p, ctx->block, SHA512_DATA_SIZE);
+	memcpy(octx.u.p, ctx->block, SHA512_BLOCK_SIZE);
 	octx.num = ctx->index;
 
-	res = length % SHA512_DATA_SIZE;
+	res = length % SHA512_BLOCK_SIZE;
 	length -= res;
 
 	if (length > 0) {
-		t2 = length / SHA512_DATA_SIZE;
+		t2 = length / SHA512_BLOCK_SIZE;
 		sha512_block_data_order(&octx, data, t2);
 
 		for (i = 0; i < t2; i++)
diff --git a/lib/accelerated/x86/x86-common.h b/lib/accelerated/x86/x86-common.h
index d91947d9c..c92c5a30c 100644
--- a/lib/accelerated/x86/x86-common.h
+++ b/lib/accelerated/x86/x86-common.h
@@ -39,7 +39,7 @@ void gnutls_cpuid(unsigned int func, unsigned int *ax, unsigned int *bx,
 	{ #name,                                      \
 	  sizeof(struct name##_ctx),                  \
 	  NAME##_DIGEST_SIZE,                         \
-	  NAME##_DATA_SIZE,                           \
+	  NAME##_BLOCK_SIZE,                          \
 	  (nettle_hash_init_func *)name##_init,       \
 	  (nettle_hash_update_func *)update_func,     \
 	  (nettle_hash_digest_func *)digest_func }
-- 
2.53.0

